src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java

```diff
@@ -54,11 +109,298 @@ private CSupport() {}
      * @throws IllegalAccessError if the runtime property {@code foreign.restricted} is not set to either
      * {@code permit}, {@code warn} or {@code debug} (the default value is set to {@code deny}).
      */
-    public static ForeignLinker getSystemLinker() {
-        Utils.checkRestrictedAccess("CSupport.getSystemLinker");
+    static CLinker getInstance() {
+        Utils.checkRestrictedAccess("CLinker.getInstance");
         return SharedUtils.getSystemLinker();
     }
 
+    /**
+     * Obtain a method handle which can be used to call a given native function.
+     *
+     * @param symbol   downcall symbol.
+     * @param type     the method type.
+     * @param function the function descriptor.
+     * @return the downcall method handle.
+     * @throws IllegalArgumentException in the case of a carrier type and memory layout mismatch.
+     */
+    MethodHandle downcallHandle(Addressable symbol, MethodType type, FunctionDescriptor function);
+
+    /**
+     * Allocates a native stub segment which contains executable code to upcall into a given method handle.
+     *
+     * <p>The base address of the returned stub segment can be passed to other foreign functions
+     * (as a function pointer). The returned segment is <em>not</em> thread-confined, and it only features
+     * the {@link MemorySegment#CLOSE} access mode. When the returned segment is closed,
+     * the corresponding native stub will be deallocated.</p>
+     *
+     * <p>The method type of the target method handle is used for linking</p>
```

Comment:
I think we can leave this out

Reply:
@PaulSandoz mentioned (and I agree) that the terms upcalls and downcalls are colloquial, in the sense that it's technically possible to pass an upcall stub to `downcallHandle`, so that it's no longer really making a downcall and, conversely, it's possible to pass a native method handle to `upcallStub` so that it will not, strictly speaking, execute an upcall to Java.

I think the real nature of what we're describing here is:

* given an address and a function descriptor, obtain a MethodHandle which can be used to execute the code at that address from Java code

* given a MethodHandle and a function descriptor, obtain a memory segment, whose base address can be passed to native code where a function pointer is expected

I don't know if the terms downcallHandle and upcallStub are the best to describe what these do.

Reply:
Maybe we can incorporate 'link' in the name? Since we're using CLinker as a name as well. Maybe we can call them: `linkForeignFunction` and `linkMethodHandle`?

Reply:
I'll save the rename of the methods for later, and integrate this so that the refactor can go through.